File: leetcode/MayChallenge/FloodFill.py
image=[[0,0,0],[0,1,0]]
sr=1
sc=1
newColor=2



col=len(image[0])
row=len(image)
visited=[[0 for c in range(col)] for r in range(row)]
_x=[0,0,1,-1]
_y=[1,-1,0,0]
q=[]
target=image[sr][sc]

#BFS
q.append([sr,sc])

while q:

    x,y=q.pop(0)
    image[x][y]=newColor

    for i in range(4):
        nx=x+_x[i]
        ny=y+_y[i]

        if 0<=nx<row and 0<=ny<col and image[nx][ny]==target and visited[nx][ny]==0:
            visited[nx][ny]=1
            q.append([nx,ny])

print(image)

File: baekjoonOnlineJudge/DFS_BFS/FloodFill.java
public class FloodFill {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
      if (image[sr][sc] == color)
        return image;

      int row = image.length;
      int col = image[0].length;

      dfs(image, sr, sc, image[sr][sc], row, col, color);

      return image;
    }

    public void dfs(int[][] image, int i, int j, int color, int row, int col, int newColor) {
      if (image[i][j] == color) {
        image[i][j] = newColor;

        if (i - 1 >= 0)
          dfs(image, i - 1, j, color, row, col, newColor);
        if (i + 1 < row)
          dfs(image, i + 1, j, color, row, col, newColor);
        if (j - 1 >= 0)
          dfs(image, i, j - 1, color, row, col, newColor);
        if (j + 1 < col)
          dfs(image, i, j + 1, color, row, col, newColor);
      }
    }
}
